### PR TITLE
Performance fine-tuning recipes for llama3 8b + 70b

### PR DESCRIPTION
**Pull request by vysarge**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/model/llama.py [hunk 1]
```diff
@@ -56,7 +56,6 @@ class LlamaConfig(GPTConfig):
     persist_layer_norm: bool = True
     bias_dropout_fusion: bool = True
     apply_rope_fusion: bool = True
-    cross_entropy_loss_fusion: bool = False
```
Comment (jiemingz) [inline nemo/collections/llm/gpt/model/llama.py:59]: wouldnt this disable this for pretraining + SFT?

Comment (vysarge) [reply nemo/collections/llm/gpt/model/llama.py:59]: Yep-- I've removed this line to allow the default cross_entropy_loss_fusion=True in GPTConfig to propagate down, while the recipes for llama3 will set it False if LoRA is used.

**Review by jiemingz — APPROVED** [review]

File: nemo/collections/llm/recipes/llama3_70b.py [hunk 2]
```diff
@@ -277,18 +284,104 @@ def finetune_recipe(
...
+            DistributedDataParallelConfig,
+            check_for_nan_in_grad=True,
+            grad_reduce_in_fp32=False,
+            overlap_grad_reduce=True,
+            overlap_param_gather=True,
+            average_in_collective=True,
+        )
+        recipe.trainer.callbacks.append(
+            run.Config(
+                MegatronCommOverlapCallback,
```
Comment (JimmyZhang12) [inline nemo/collections/llm/recipes/llama3_70b.py:368]: can you add wgrad deferral for SFT?

File: nemo/collections/llm/recipes/llama3_70b.py [hunk 3]
```diff
@@ -277,18 +284,104 @@ def finetune_recipe(
...
     elif peft_scheme.lower() == 'lora':
         recipe.peft = run.Config(LoRA)
+        recipe.peft.dim = 16
+        recipe.peft.alpha = 32
+        recipe.peft.target_modules = ['linear_qkv']
+
+        # some settings currently do not function correctly with LoRA
+        recipe.model.config.cross_entropy_loss_fusion = False
+        recipe.trainer.strategy.ckpt_async_save = False
+        recipe.trainer.strategy.ddp = "megatron"
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/llama3_70b.py:309]: not big problems but
`ckpt_async_save` works after #10913 now,
and ddp strategy is by default "megatron" in default_finetune_recipe

File: nemo/collections/llm/recipes/llama3_8b.py [hunk 4]
```diff
@@ -273,7 +285,82 @@ def finetune_recipe(
         recipe.optim.config.lr = 5e-6
     elif peft_scheme.lower() == 'lora':
         recipe.peft = run.Config(LoRA)
+        recipe.peft.dim = 8
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/llama3_8b.py:288]: have you verified that dim=8 would not compromise accuracy?

File: nemo/collections/llm/recipes/llama3_8b.py [hunk 5]
```diff
@@ -273,7 +285,82 @@ def finetune_recipe(
...
+        peft_scheme (str): Name of the peft scheme to use for fine-tuning. Allowed values: 'lora', 'none'/None.
+
+    Returns:
+        run.Partial: Partial configuration for performance-optimized fine-tuning.
+
+    Note:
+        Use this method with caution and only when you need maximum performance.
+        It may not be suitable for all hardware configurations or use cases.
+    """
+    recipe.trainer.strategy.tensor_model_parallel_size = 1
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/llama3_8b.py:335]: just checking, TP1 works for full finetuning as well?

**Review by cuichenx — COMMENTED** [review]
Just a few small questions

File: nemo/collections/llm/recipes/llama3_70b.py [hunk 6]
```diff
@@ -277,18 +284,104 @@ def finetune_recipe(
...
+
         recipe.trainer.strategy.tensor_model_parallel_size = 8
         recipe.optim.config.lr = 1e-4
     else:
         raise ValueError(f"Unrecognized peft scheme: {peft_scheme}")
+
+    # Sequence length settings in the model and dataset must agree
+    recipe.model.config.seq_length = seq_length
+    recipe.data.seq_length = seq_length
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/llama3_70b.py:318]: without packed sequence, we usually use seq_length=2048 (because most samples in Squad are well shorter than 2048 tokens)

File: nemo/collections/llm/recipes/llama3_8b.py (hunk 5 above)
Comment (vysarge) [reply nemo/collections/llm/recipes/llama3_8b.py:335]: With BF16 grad it should.  Let me double-check.

File: nemo/collections/llm/recipes/llama3_70b.py (hunk 2 above)
Comment (vysarge) [reply nemo/collections/llm/recipes/llama3_70b.py:371]: done.

Comment (JimmyZhang12) [reply nemo/collections/llm/recipes/llama3_70b.py:371]: to be safe can you also add:
`wgrad_deferral_limit=22`

File: nemo/collections/llm/recipes/llama3_70b.py (hunk 3 above)
Comment (vysarge) [reply nemo/collections/llm/recipes/llama3_70b.py]: thanks, removed the unneeded settings!

File: nemo/collections/llm/recipes/llama3_8b.py (hunk 5 above)
Comment (vysarge) [reply nemo/collections/llm/recipes/llama3_8b.py:337]: Yep, both fit.

File: nemo/collections/llm/recipes/llama3_70b.py (hunk 6 above)
Comment (vysarge) [reply nemo/collections/llm/recipes/llama3_70b.py:320]: Just altered this to use 4K for packed and 2K for unpacked by default.

**Review by cuichenx — DISMISSED** [review]
LGTM

File: nemo/collections/llm/recipes/llama3_70b.py (hunk 2 above)
Comment (vysarge) [reply nemo/collections/llm/recipes/llama3_70b.py:371]: Do we need this setting also for pretrain?  I see it sets defer_embedding_wgrad_compute=True as well.

Comment (JimmyZhang12) [reply nemo/collections/llm/recipes/llama3_70b.py:371]: https://github.com/NVIDIA/NeMo/blob/c5b39e84774846aa4907dcb7e9ff70a3022ce400/nemo/collections/llm/recipes/llama3_70b.py#L234
Yes pretrain sets it here

**Review by JimmyZhang12 — APPROVED** [review]
